File: code/src/main/java/edu/ncsu/haskell/parser/AntlrParser.java
package edu.ncsu.haskell.parser;

import edu.ncsu.haskell.parser.antlr.HaskellLexer;
import edu.ncsu.haskell.parser.antlr.HaskellParser;
import org.antlr.v4.runtime.*;
import org.antlr.v4.runtime.misc.Predicate;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNodeImpl;
import org.antlr.v4.runtime.tree.Tree;
import org.antlr.v4.runtime.tree.Trees;

import java.util.Arrays;
import java.util.List;


class FunctionNodePredicate implements Predicate<ParseTree> {
    @Override
    public boolean test(ParseTree tree) {
        if (tree instanceof TerminalNodeImpl) {
            return false;
        }
        String name = TreeUtils.getAntlrNodeName(tree);
        return name.toLowerCase().equals("decl_no_th");
    }
}

public class AntlrParser {



    public static void main(String[] args) {
        String haskellCode = "in_range min max x = x >= min && x <= max";
        CharStream input = CharStreams.fromString(haskellCode);
        HaskellLexer lexer = new HaskellLexer(input);
        CommonTokenStream tokens = new CommonTokenStream(lexer);
        HaskellParser parser = new HaskellParser(tokens);
        RuleContext tree = parser.module();
        Predicate p = new FunctionNodePredicate();
        System.out.println(TreeUtils.toJson(tree));
        Tree funcTree = Trees.findNodeSuchThat(tree, p);
        System.out.println(TreeUtils.toJson(funcTree));


//        List<String> ruleNamesList = Arrays.asList(parser.getRuleNames());
//        String parseTree = TreeUtils.toPrettyTree(tree, ruleNamesList);
//        System.out.println(parseTree);
    }

}

File: code/src/main/python/haskell/parsers/antlr_parser.py
import sys
import os

sys.path.append(os.path.abspath("."))
sys.dont_write_bytecode = True

__author__ = "COSAL"

from antlr4 import *
from haskell.parsers.antlr.HaskellParser import HaskellParser
from haskell.parsers.antlr.HaskellLexer import HaskellLexer
from haskell.parsers.antlr.HaskellParserListener import HaskellParserListener


txt = """
in_range min max x = x >= min && x <= max
"""

def test():
  lexer = HaskellLexer(txt)
  stream = CommonTokenStream(lexer)
  parser = HaskellParser(stream)
  tree = parser.module()
  print(tree.toStringTree(parser))


if __name__ == "__main__":
  test()
